cmd/gcsim/optimize_substats.go

```diff
@@ -0,0 +1,673 @@
+	artifactSets4Star = []string{
+		"instructor",
+	}
+
+	substatValues  = make([]float64, core.EndStatType)
+	mainstatValues = make([]float64, core.EndStatType)
+)
+
+// Additional runtime option to optimize substats according to KQM standards
+func runSubstatOptim(o opts) {
+	// Substat Optimization strategy is very simplistic right now:
+	// This is not fully optimal - see other comments in code
+	// 1) User sets team, weapons, artifact sets/main stats, and rotation
+	// 2) Given those, for each character, sim picks ER substat value that functionally maximizes DPS Mean/SD,
+	// subject to a penalty on high ER values
+	//    - Strategy is to just do a dumb grid search over ER substat values for each character
+	//    - ER substat values are set in increments of 2 to make the search easier
+	// 3) Given ER values, we then optimize the other substats by doing a "gradient descent" (but not really) method
+
+	// TODO: Is this actually the best way to set these values or am I missing something..?
+	substatValues[core.ATKP] = 0.0496
+	substatValues[core.CR] = 0.0331
+	substatValues[core.CD] = 0.0662
+	substatValues[core.EM] = 19.82
+	substatValues[core.ER] = 0.0551
+	substatValues[core.HPP] = 0.0496
+	substatValues[core.DEFP] = 0.062
+	substatValues[core.ATK] = 16.54
+	substatValues[core.DEF] = 19.68
+	substatValues[core.HP] = 253.94
+
+	// Used to try to back out artifact main stats for limits
+	// TODO: Not sure how to handle 4* artifact sets... Config can't really identify these instances easily
+	// Most people will have 1 5* artifact which messes things up
+	mainstatValues[core.ATKP] = 0.466
+	mainstatValues[core.CR] = 0.311
+	mainstatValues[core.CD] = 0.622
+	mainstatValues[core.EM] = 186.5
+	mainstatValues[core.ER] = 0.518
+	mainstatValues[core.HPP] = 0.466
+	mainstatValues[core.DEFP] = 0.583
+
+	// Start logger
+	zapcfg := zap.NewDevelopmentConfig()
+	zapcfg.Level = zap.NewAtomicLevelAt(zapcore.InfoLevel)
+	zapcfg.EncoderConfig.CallerKey = ""
+	zapcfg.EncoderConfig.StacktraceKey = ""
+	zapcfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+
+	if o.verbose {
+		zapcfg.Level = zap.NewAtomicLevelAt(zapcore.DebugLevel)
+	}
+	logger, _ := zapcfg.Build()
+	defer logger.Sync()
+	sugarLog := logger.Sugar()
+
+	// Parse config
+	src, err := ioutil.ReadFile(o.config)
+	if err != nil {
+		sugarLog.Error(err)
+		os.Exit(1)
+	}
+
+	//check for imports
+	var data strings.Builder
+	var re = regexp.MustCompile(`(?m)^import "(.+)"$`)
+
+	// Regex to identify main stats based on flower. Check that characters all have one that we can recognize
+	var reMainstats = regexp.MustCompile(`(?m)^[A-z]+ add stats.* hp=(4780|3571).*\n`)
+	if len(reMainstats.FindAll(src, -1)) != 4 {
+		sugarLog.Error("Error: Could not identify valid main artifact stat rows for all characters based on flower HP values.")
+		sugarLog.Error("5* flowers must have 4780 HP, and 4* flowers must have 3571 HP.")
+		os.Exit(1)
+	}
```

Comment:
```suggestion
	var reMainstats = regexp.MustCompile(`(?m)^[A-z]+ add stats.* hp=(4780|3571).*\n`)
	var numChars = regexp.MustCompile(`char`)

	if len(reMainstats.FindAll(src, -1)) != len(numChars.FindAll(src, -1)) {
		sugarLog.Error("Error: Could not identify valid main artifact stat rows for all characters based on flower HP values.")
		sugarLog.Error("5* flowers must have 4780 HP, and 4* flowers must have 3571 HP.")
		os.Exit(1)
	}

```

To allow for substat optimization for less than 4 chars